src/parallel_parameter_search/walk_optimization.py

```diff
@@ -0,0 +1,508 @@
+            self.sim.set_joints(joint_command)
+            self.last_time = current_time
+
+    def complete_walking_step(self, number_steps=1, fake=False):
+        start_time = self.sim.get_time()
+        for i in range(number_steps):
+            # does a double step
+            while not rospy.is_shutdown():
+                current_time = self.sim.get_time()
+                joint_command = self.walk.step(current_time - self.last_time, self.current_speed,
+                                               self.sim.get_imu_msg(),
+                                               self.sim.get_joint_state_msg(), self.sim.get_pressure_left(),
+                                               self.sim.get_pressure_right())
+                self.sim.set_joints(joint_command)
+                self.last_time = current_time
+                if not fake:
+                    self.sim.step_sim()
+                phase = self.walk.get_phase()
+                # phase will advance by the simulation time times walk frequency
+                next_phase = phase + self.sim.get_timestep() * self.walk.get_freq()
+                # do double step to always have torso at same position
+                if (phase >= 0.5 and next_phase >= 1.0):  # or (phase < 0.5 and next_phase >= 0.5):
+                    # next time the walking step will change
+                    break
+                if self.sim.get_time() - start_time > 5:
+                    # if walking does not perform step correctly due to impossible IK problems, do a timeout
+                    break
+
+    def compute_cost(self, x, y, yaw):
+        """
+        x,y,yaw have to be either 1 for being goo, -1 for being bad or 0 for making no difference
+        """
+        # factor to increase the weight of the yaw since it is a different unit then x and y
+        yaw_factor = 5
+        # todo better formular
+        pos, rpy = self.sim.get_robot_pose_rpy()
+
+        # 2D pose
+        current_pose = [pos[0], pos[1], rpy[2]]
+        correct_pose = [x * self.time_limit,
+                        y * self.time_limit,
+                        (yaw * self.time_limit) % (2 * math.pi)]  # todo we dont take multiple rotations into account
+        cost = abs(current_pose[0] - correct_pose[0]) \
+               + abs(current_pose[1] - correct_pose[1]) \
+               + abs(current_pose[2] - correct_pose[
+            2]) * yaw_factor  # todo take closest distance in circle through 0 into account
+        # method doesn't work for going forward and turning at the same times
+        if yaw != 0:  # and (x != 0 or y != 0):
+            # just give 0 cost for surviving
+            cost = 0
+        # test if robot moved at all for simple case
+        early_term = False
+        if yaw == 0 and ((x != 0 and abs(current_pose[0]) < 0.5 * abs(correct_pose[0])) or (
+                y != 0 and abs(current_pose[1]) < 0.5 * abs(correct_pose[1]))):
+            early_term = True
+            print("didn't move")
```

Comment:
Should the debug print be kept?

Reply:
yes, makes it more clear why the optimization is stopped at this point